// @veams/plugin-store/lib/index.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Imports
 */
var subject_1 = require("./subject");
exports.Subject = subject_1.default;
/**
 * Internal data management
 */
var store;
exports.store = store;
var internalState = {};
exports.state = internalState;
var internalReducer = function (action, payload) {
};
var internalSubjects = {};
/**
 * Broadcast functions
 */
function broadcastAll(subjects) {
    if (subjects === void 0) { subjects = internalSubjects; }
    Object.keys(subjects).forEach(function (subject) {
        subjects[subject].next(internalState[subject]);
    });
}
exports.broadcastAll = broadcastAll;
function broadcast(subject) {
    internalSubjects[subject].next(internalState[subject]);
}
exports.broadcast = broadcast;
/**
 * Configure store
 */
function configure(reducer, state, subjects) {
    exports.state = internalState = state;
    internalReducer = reducer;
    internalSubjects = subjects;
}
/**
 * Internal store class
 */
var AppStore = /** @class */ (function () {
    function AppStore(reducer, state, subjects) {
        configure(reducer, state, subjects);
        console.log('Store initialized!');
    }
    AppStore.prototype.subscribe = function (observer) {
        Object.keys(internalSubjects).forEach(function (subject) {
            internalSubjects[subject].subscribe(observer);
            observer.next(internalState[subject]);
        });
    };
    AppStore.prototype.unsubscribe = function (observer) {
        Object.keys(internalSubjects).forEach(function (subject) {
            internalSubjects[subject].unscribe(observer);
        });
    };
    AppStore.prototype.select = function (subtype) {
        return internalSubjects[subtype];
    };
    AppStore.prototype.dispatch = function (action, payload) {
        internalReducer(action, payload);
    };
    return AppStore;
}());
/**
 * Plugin
 */
var VeamsStore = {
    pluginName: 'Store',
    initialize: function (Veams, _a) {
        var reducer = _a.reducer, state = _a.state, subjects = _a.subjects;
        exports.store = store = new AppStore(reducer, state, subjects);
    }
};
exports.default = VeamsStore;
